keps/sig-network/2086-service-internal-traffic-policy/README.md

```diff
@@ -207,54 +192,55 @@ Yes, the feature gate can be disabled, but Service resource that have set the ne
 
 * **What happens if we reenable the feature if it was previously rolled back?**
 
-New Services should be able to set the `trafficPolicy` field. Existing Services that have the field set already should not be impacted.
+New Services should be able to set the `internalTrafficPolicy` field. Existing Services that have the field set already should not be impacted.
 
 * **Are there any tests for feature enablement/disablement?**
 
-There will be unit tests to verify that apiserver will drop the field when the `ServiceTrafficPolicy` feature gate is disabled.
+There will be unit tests to verify that apiserver will drop the field when the `ServiceInternalTrafficPolicy` feature gate is disabled.
 
 ### Rollout, Upgrade and Rollback Planning
 
 _This section must be completed when targeting beta graduation to a release._
 
 * **How can a rollout fail? Can it impact already running workloads?**
 
-TBD for beta.
+Rollout should have minimal impact because the default value of `internalTrafficPolicy` is `Cluster`, which is the default behavior today.
 
 * **What specific metrics should inform a rollback?**
 
-TBD for beta.
+N/A
 
 * **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
 
-TBD for beta.
+No, but this will be manually tested prior to beta. Automated testing will be done if the test tooling is available.
 
 * **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
 fields of API types, flags, etc.?**
 
-TBD for beta.
+No.
 
 ### Monitoring Requirements
 
 _This section must be completed when targeting beta graduation to a release._
 
 * **How can an operator determine if the feature is in use by workloads?**
 
-TBD for beta.
+* Check Service to see if `internalTrafficPolicy` is set to `Local`.
+* Check iptable rules on kube-proxy to see if cluster IP routing is only being sent to node-local endpoints.
```

Comment:
Operators can't reliably do that if they manage hundreds or thousands of clusters.

Can we add a metric to kube-proxy that will be exposing it (e.g. some counter that incremenets when we add node-local endpoints or sth like that?)

Reply:
I could, but there's not much precedent for this. If we did it for internal traffic policy, we'd want to do it for a bunch of other error cases with kube-proxy. Do we really want that?

Reply:
@thockin - thoughs?

I would say that the fact that we're missing observability now doesn't mean that we should continue keeping it non-observable. So I would say that let's start adding for new stuff at least.

Reply:
It seems reasonable that we could add metrics for kube-proxy, to capture its perceived breakdown of things.  We have not much to build on here, though.

I could imagine a metric that simply tracks the number of services labelled by external and internal traffic policy, but that doesn't say much about this feature.  We could add a metric that counts local vs non-local endpoints but that's not useful unless we carve it up by internal and external traffic policy.  I'm worried that the logic to carve up the metrics becomes about as complicated as the logic it is tracking.

What are we really trying to answer?  Something like "are cases where policy is local staying local?", right?  We can't expose per-service metrics, and absolute numbers are not super useful.  I'm skeptical that the code would even have a path for recognizing "local policy, non-local endpoint" to be counted.

What about something simple like "blackholes" labelled by ITP ?  If a service has no endpoints, that is interesting.  If a service has ITP=local and no endpoints, that's alertable (probably).

intersecting that with the other kep, maybe label by ITP and by conditions?

Reply:
> If a service has ITP=local and no endpoints, that's alertable (probably)

Not too familiar with all the use cases here, but one I can imagine would involve different sets of nodes for different workloads. In that case traffic from a needs to reach b where ITP=local, but both a and b only run on the same subset of nodes. Technically that would result in a lot of blackholes for other nodes where the Service existed but local endpoints did not. Not sure how big of a concern that is or if it meaningfully limits the usefulness of that metric, it's just one of the ways I could imagine ITP being used.

Reply:
I agree it's possible, and the good news is we don't have to define alerts here.  :)  My point stands - when it comes to traffic policy "working" we want to answer 2 questions:

1) Are we finding the local endpoints? (blackholes = 0)
2) Are we ONLY finding the local endpoints?

The 2nd is much harder to instrument without splitting the impl up, I think.  The code path that would accumulate non-local endpoints for a local service doesn't exist because it is non-sense, and we don't carry an intermediate representation that we can cross-check ex post facto (which we can and maybe should fix),

E.g.  convert the sync loop to 2 stages.  First, loop over all services and gather the endpoints for each "path" (internal/external client).  Second, convert that into iptables/ipvs.  Then you can run sanity checks in between like "if service is ITP local and has non-local endpoints, there's a bug".  

Unsurprisingly, this sounds quite a lot like kpng.  So maybe adding blackholes is a good start and we make sure kpng can handle such other metrics?

@mcluseau

Reply:
Great discussion. And I buy the argument that doing (2) would be hard and the logic would be similarly complex to the code itself.

So if we can just export metric about blackholes - that sounds fine to me (and that seems more important imho).
So yeah - let's just expose blackholes.

Reply:
Added some words to add the "blackhole" metrics, we can discuss naming of the actual metric during PR review